### PR TITLE
Feature: Adding support for Vindex Hints to allow for greater control over shard routing

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description
Implemented support for vindex hints in Vitess, allowing for explicit shard routing control. This feature introduces `USE VINDEX` and `IGNORE VINDEX` hints, enhancing query performance by guiding the query planner's routing decisions. See the linked RFC for full details.

## Related Issue(s)
Implements RFC #15171
Website docs: https://github.com/vitessio/website/pull/1683

## Checklist
-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required - https://github.com/vitessio/website/pull/1683

### CODE REVIEW

File: go/vt/vtgate/semantics/early_rewriter.go [hunk 1]
```diff
@@ -140,30 +140,47 @@ func rewriteNotExpr(cursor *sqlparser.Cursor, node *sqlparser.NotExpr) {
...
+				return true
+			}
+
+			err = innerErr
+			return false
+		})
+		return err
+	case *sqlparser.AliasedTableExpr:
+		// this rewriting is done in the `up` phase, because we need the vindex hints to have been
+		// processed while collecting the tables.
```
[the remark this answers is not in the document]
Comment (GuptaManan100) [reply go/vt/vtgate/semantics/early_rewriter.go:170]: Fixed

File: go/vt/vtgate/semantics/early_rewriter.go [hunk 2]
```diff
@@ -140,30 +140,53 @@ func rewriteNotExpr(cursor *sqlparser.Cursor, node *sqlparser.NotExpr) {
...
 func (r *earlyRewriter) up(cursor *sqlparser.Cursor) error {
-	// this rewriting is done in the `up` phase, because we need the scope to have been
-	// filled in with the available tables
-	node, ok := cursor.Node().(*sqlparser.JoinTableExpr)
-	if !ok || len(node.Condition.Using) == 0 {
-		return nil
-	}
+	switch node := cursor.Node().(type) {
+	case *sqlparser.JoinTableExpr:
+		// this rewriting is done in the `up` phase, because we need the scope to have been
```
Comment (systay) [inline go/vt/vtgate/semantics/early_rewriter.go:145]: nitpick: might as well extract this into it's own method

**Comment by systay** [comment]
In the discussion of the RFC, we talked about doing the same thing MySQL does, and fail for vindexes that do not exist. We should add a test or two that checks this scenario

**Review by systay — APPROVED** [review]

**Review by GuptaManan100 — COMMENTED** [review]
I can't approve the PR, but I would approve the changes that @systay made if I could.

File: go/vt/vterrors/code.go [hunk 3]
```diff
@@ -89,6 +89,8 @@ var (
 	VT09017 = errorWithoutState("VT09017", vtrpcpb.Code_FAILED_PRECONDITION, "%s", "Invalid syntax for the statement type.")
 	VT09018 = errorWithoutState("VT09018", vtrpcpb.Code_FAILED_PRECONDITION, "%s", "Invalid syntax for the vindex function statement.")
 	VT09019 = errorWithoutState("VT09019", vtrpcpb.Code_FAILED_PRECONDITION, "keyspace '%s' has cyclic foreign keys", "Vitess doesn't support cyclic foreign keys.")
+	VT09020 = errorWithoutState("VT09020", vtrpcpb.Code_FAILED_PRECONDITION, "can not use multiple vindex hints for table %s", "Vitess does not allow using multiple vindex hints on the same table.")
+	VT09021 = errorWithoutState("VT09021", vtrpcpb.Code_FAILED_PRECONDITION, "no such vindex found: %s", "Vindex hints have to reference an existing vindex, and no such vindex could be found.")
```
Comment (harshit-gangal) [inline go/vt/vterrors/code.go:93]: should we use the same error code as mysql
```
ERROR 1176 (42000): Vindex '<vindex name>' doesn't exist in table '<table name>'
```

File: go/vt/vterrors/code.go [hunk 4]
```diff
@@ -89,6 +89,8 @@ var (
 	VT09017 = errorWithoutState("VT09017", vtrpcpb.Code_FAILED_PRECONDITION, "%s", "Invalid syntax for the statement type.")
 	VT09018 = errorWithoutState("VT09018", vtrpcpb.Code_FAILED_PRECONDITION, "%s", "Invalid syntax for the vindex function statement.")
 	VT09019 = errorWithoutState("VT09019", vtrpcpb.Code_FAILED_PRECONDITION, "keyspace '%s' has cyclic foreign keys", "Vitess doesn't support cyclic foreign keys.")
+	VT09020 = errorWithoutState("VT09020", vtrpcpb.Code_FAILED_PRECONDITION, "can not use multiple vindex hints for table %s", "Vitess does not allow using multiple vindex hints on the same table.")
```
Comment (harshit-gangal) [inline go/vt/vterrors/code.go:92]: this is more like an unsupported error. But, as it is not part of MySQL standard. This type of error should be fine.

File: go/vt/vtgate/planbuilder/operators/sharded_routing.go [hunk 5]
```diff
@@ -87,6 +113,16 @@ func newShardedRouting(vtable *vindexes.Table, id semantics.TableSet) Routing {
...
+// indexesContains is a helper function that returns whether a given string is part of the IdentifierCI list.
+func indexesContains(indexes []sqlparser.IdentifierCI, name string) bool {
+	for _, index := range indexes {
+		if index.EqualString(name) {
+			return true
+		}
+	}
+	return false
+}
+
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/operators/sharded_routing.go:125]: nit: could use `slices.IndexFunc`

**Review by harshit-gangal — COMMENTED** [review]
I think these changes should seamlessly work for the join query as well. Can you add a test case for them as well?

**Pull request merged**